pkg/injector/bootstrap.go

```diff
@@ -100,3 +104,116 @@ func (wh *mutatingWebhook) marshalAndSaveBootstrap(name, namespace string, confi
+		}
+
+		if !b.certManager.ShouldRotate(bootstrapCert) {
+			continue
+		}
+
+		issuedCert, err := b.certManager.IssueCertificate(bootstrapCert.CommonName.String(), certificate.Internal, certificate.FullCNProvided())
+		if err != nil {
+			log.Error().Err(err).Str(errcode.Kind, errcode.GetErrCodeWithMetric(errcode.ErrRotatingCert)).Msgf("Error rotating cert %s", issuedCert)
+		}
+
+		secret.Data[bootstrap.EnvoyXDSCACertFile] = issuedCert.GetTrustedCAs()
+		secret.Data[bootstrap.EnvoyXDSCertFile] = issuedCert.GetCertificateChain()
+		secret.Data[bootstrap.EnvoyXDSKeyFile] = issuedCert.GetPrivateKey()
+
+		err = b.kubeController.UpdateSecret(ctx, secret, secret.Data)
+		if err != nil {
+			if apierrors.IsConflict(err) {
+				log.Error().Err(err).Str(errcode.Kind, errcode.GetErrCodeWithMetric(errcode.ErrUpdatingBootstrapSecret)).
+					Msgf("There was an update conflict while trying to update the envoy bootstrap config secret %s/%s with issued cert %s", secret.Namespace, secret.Name, issuedCert)
+				continue
+			}
+			log.Error().Err(err).Str(errcode.Kind, errcode.GetErrCodeWithMetric(errcode.ErrUpdatingBootstrapSecret)).
+				Msgf("Error updating bootstrap secret %s/%s with issued cert %s", secret.Namespace, secret.Name, issuedCert)
+			continue
+		}
+	}
+}
+
+func getCert(secret *corev1.Secret) (*certificate.Certificate, error) {
+	pemCert, ok := secret.Data[bootstrap.EnvoyXDSCertFile]
+	if !ok {
+		// TODO(#3962): metric might not be scraped before process restart resulting from this error
+		log.Error().Err(certificate.ErrInvalidCertSecret).Str(errcode.Kind, errcode.GetErrCodeWithMetric(errcode.ErrObtainingCertFromSecret)).
+			Msgf("Opaque k8s secret %s/%s does not have required field %q", secret.Namespace, secret.Name, bootstrap.EnvoyXDSCertFile)
+		return nil, certificate.ErrInvalidCertSecret
+	}
+
+	pemKey, ok := secret.Data[bootstrap.EnvoyXDSKeyFile]
+	if !ok {
+		// TODO(#3962): metric might not be scraped before process restart resulting from this error
+		log.Error().Err(certificate.ErrInvalidCertSecret).Str(errcode.Kind, errcode.GetErrCodeWithMetric(errcode.ErrObtainingPrivateKeyFromSecret)).
+			Msgf("Opaque k8s secret %s/%s does not have required field %q", secret.Namespace, secret.Name, bootstrap.EnvoyXDSKeyFile)
+		return nil, certificate.ErrInvalidCertSecret
+	}
+
+	x509Cert, err := certificate.DecodePEMCertificate(pemCert)
+	if err != nil {
+		// TODO(#3962): metric might not be scraped before process restart resulting from this error
+		log.Error().Err(err).Str(errcode.Kind, errcode.GetErrCodeWithMetric(errcode.ErrDecodingPEMCert)).
+			Msg("Error converting PEM cert to x509 to obtain serial number")
+		return nil, err
+	}
+
+	return &certificate.Certificate{
+		CommonName:   certificate.CommonName(x509Cert.Subject.CommonName),
+		SerialNumber: certificate.SerialNumber(x509Cert.SerialNumber.String()),
+		CertChain:    pemCert,
+		TrustedCAs:   secret.Data[bootstrap.EnvoyXDSCACertFile],
```

Comment:
nit: We are not verifying that the secret has the ca cert file field as we do for the key file and pem cert fields above. Not sure if this check is necessary.